### DISCUSSION TITLE
Snake-like Robot Modeling and Navigation (Project 224)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#56] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project 224](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Snake-like%20Robot%20Modeling%20and%20Navigation)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by Nachiket497** [comment]
Hello, my name is Nachiket Sawwalakhe and I am a student at VNIT Nagpur. I am interested in working and collaborating on this project.

**Comment by SafalRaj10** [comment]
Hi, This is Safal Raj, a Mechanical Engineering student at IIT Patna. I'm intrigued and will be contributing to this project.

**Comment by grsasaki-tmw** [comment]
Hello @Nachiket497 and @SafalRaj10, we do appreciate your interest in this project! If you have any questions or points you would like to discuss, I welcome and encourage you to post here.  Best Regards, Gen Sasaki

**Comment by Nachiket497** [comment]
I and @amit-420  have created a simulink model a snake and we want to run it on ground plane. How to add a ground env along with the snake model ? can you please provide some resources on this?
@grsasaki-tmw

**Reply by grsasaki-tmw** [reply]
@Nachiket497, assuming you are using Simscape Multibody as the project recommends, you can use the Infinite Plane block to define the ground plane. (https://www.mathworks.com/help/sm/ref/infiniteplane.html)

**Reply by Nachiket497** [reply]
Will check on this.

**Reply by amit-420** [reply]
@grsasaki-tmw we are trying to implement infiniteplane in snake model. As of now we have connected "R" of the block to world frame and "G"  to the contact force block as mentioned in the documentation but we cannot find how to move ahead. Could you provide any example where infinte plane is used with a CAD model with multiple contact points ?

**Comment by bhavikpatel2** [comment]
Hi, I am a research scholar at IIT Guwahati and am interested in working and collaborating on this project

**Reply by bhavikpatel2** [reply]
I have created a Simulink model of the Snake robot. The one end is permanently fixed with the world frame. I tried a 6DOF joint between the world frame and the link joint, but it doesn't work. Next, I tried a planar joint, but it also failed. Could you suggest to me how the last link can move?@Nachiket497

**Reply by robertogl** [reply]
Hello @bhavikpatel2,

I do not think this is a joint problem. If my understanding of your problem is correct you may need to add a transform between the joint and the world frame which value will depend on the robot's motion. The block is called [Transform Sensor](https://www.mathworks.com/help/sm/ref/transformsensor.html). 
I hope this helps.

**Reply by bhavikpatel2** [reply]
Thank you for your reply @robertogl.

**Reply by bhavikpatel2** [reply]
[Snake_Robot.zip](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/files/10131256/Snake_Robot.zip)
Hello @robertogl,

I tried as per your suggestion. Unfortunately, it failed. I have attached my file with it.

**Reply by robertogl** [reply]
Hi @bhavikpatel2 ,

I can take a look, but rather than exchanging zip files, would you mind creating a GitHub repo? It would be much easier to collaborate and propose new edits.
Thanks.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
I am new to GitHub. I don't have any problem creating a GitHub repo.

**Reply by robertogl** [reply]
Thanks! Please send the link to the repo once is available.

**Reply by bhavikpatel2** [reply]
Hello @robertogl 
Please find the GitHub repo link.
https://github.com/bhavikpatel2/Snake-robot-MATLAB.git

**Reply by bhavikpatel2** [reply]
Hello @robertogl 
Have you checked that problem?

**Reply by robertogl** [reply]
Hi @bhavikpatel2 

I apologize for the late response but we have had too many end of the year commitments. I took a very quick look at the model and although I will need more time to discover and fix the problem I suspect there might be something to do with the contact forces.

I will get back to it soon, I hope you do not have any time constraint.

**Reply by bhavikpatel2** [reply]
No problem @robertogl

**Reply by robertogl** [reply]
Hi @bhavikpatel2 

I hope you are still working on this project.

I finally checked your model in detail and noticed that you were providing motion (angles) to the joints rather than torque. In that case you provide directly the output to the joints and all you get is a visualization of your sin wave, but all the dynamics is bypassed. 
So, I changed each joint block to accept torque values and calculate the relative output motion automatically. Once done that I needed the torque to be such to generate the desired sin wave motion of the joint. That is why I added a control block (based on a simple PID) to generate the command torque that is able to follow the desired input motion (based on your previous input). 
It now seems to work fine although the snake is still not moving forward. I think for that you need a different signal input to each joint rather than a sin wave only. 

You should be able to see my changes in a pull request that I have just created to push the new model into your GitHub repository. Check the changes and if they look fine you can just merge them into your main branch. 

I hope this helps.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 

Thank you so much for your reply. 
As per my literature review, the snake robot would get motion due to friction. The sine wave is the easiest way to synchronize the serpentine motion. Here, there is a chance of two possibilities. 
1. The friction between the snake robot body and the world body is not sufficient to get motion. However, I have tried to solve that problem by considering the contact friction block. The whole geometry could not come to friction therefore at each corner, the spherical solid ball is distributed. 
2. The other possibility is the kinematic constraint. That means somewhere one joint is fixed around the reference frame.

**Reply by robertogl** [reply]
Yes, I see your serpentine signal makes sense. We might need to add an offset later on to change the direction of motion. 

Regarding the lack of translation I would try first with your option 1, friction.

Also, have you merged my changes? You need torque input to the joints.

**Reply by bhavikpatel2** [reply]
Yes, I have checked that and modified it. It is fine.

**Reply by robertogl** [reply]
Ok, perhaps you are not actively using the GitHub repo you shared as I do not see any changes there. The easiest way to check in the changes is by merging the pull request. 

Also, I found this paper as another useful resource of information about snake robot locomotion: 
https://ieeexplore.ieee.org/document/9921291

**Reply by bhavikpatel2** [reply]
Hi,

I have done it but the snake robot does not stay on the surface after the inclusion of the friction model externally.

**Reply by robertogl** [reply]
I saw that, but after checking the model it seems that behavior is due to your new friction model. Do you have any reference for that model?

**Reply by bhavikpatel2** [reply]
https://www.sciencedirect.com/science/article/pii/S1367578817301050

I follow this reference for the dynamic modelling and control of the snake robot.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 

In the simulation, I observed that the small spherical balls attached to the bottom of the surface, are touched with the ground initially but when the simulation started they went down, and the robot moved in the same place. I tried to apply the normal force externally, the robot flew away from the ground. How can we keep the balls constantly touched with the ground? That may solve the motion problem.

**Reply by robertogl** [reply]
Hi @bhavikpatel2,

It might be a transformation issue as the rigid transform between the body and the plane. I will try to take a look asap.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
I have updated the model. In this, the contact friction force is removed, and directly the forces are given from the calculation of the friction force as per the literature. The snake robot's vertical motion is now restricted. I observed the side-winding motion of the snake robot. But I feel the contact friction force has more impact than mathematically calculated friction force input because the contact between the robot body and surface can change the motion significantly.

**Reply by robertogl** [reply]
Hi @bhavikpatel2,

Sorry about the late response. Great to know that one problem has been solved. I will try to take a look within the next few days.

One thing I had not mentioned in the past (sorry about that) is using the Point Cloud block to model the geometry of each link of the snake that makes contact with the ground. Each link would have a point cloud associated with it, and then that point cloud has a contact force block between it and the ground. And then I would tune the parameters appropriately. This may probably work better than just using a sphere. Here an example:
[https://www.mathworks.com/help//sm/ug/grid_surface_ball.html](https://www.mathworks.com/help/sm/ug/grid_surface_ball.html)

Cheers!

**Reply by bhavikpatel2** [reply]
Hi @robertogl 

Thank you so much for sharing. I will look into this.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 

I got the motion of the snake robot. It moves in a straight line in the positive XY axis. Please look into this. Will it be sufficient for the submission because my extended deadline is near?

**Reply by robertogl** [reply]
Hi @bhavikpatel2,

That is awesome! Great progress and thanks for letting me know..

Would you be willing to add a direction control block so that the direction of the snake can be easily changed by a command?
Do not worry about the deadline. Up to you.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
Definitely, I will add direction control for it.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 

I have worked on the direction control of the snake robot. I have used the method for direction tracking from my previous [published work](https://journals.sagepub.com/doi/abs/10.1177/09596518231153253) for the direction tracking. However, it didn't work because the robot couldn't move in the second and fourth quadrants. The robot can move in the first (forward motion) and third (back motion) quadrants. I have tried to move it by manually direction control by constant block. However, it moves only in the first quadrant. I couldn't understand why this is happening?

**Reply by robertogl** [reply]
Hi @bhavikpatel2,

Can you provide some more details? Can the snake change its direction at all even if slightly? What happens precisely, the robot changes its direction until it stops moving?

**Reply by bhavikpatel2** [reply]
Hi @robertogl 

Generally, to generate the serpentine motion (A sin(wt+delta*b)+ phi) is used.

1. The sine wave (A sin (wt+b)) is used for the generation of serpentine motion with ( using w) speed control. The (phi) offset angle works as the direction control, like turning and it is applied for the head link and other links should follow it because of the geometric constraints. These are inputted in the local coordinate system, but we see the snake robot motion in XY frame, which is a global coordinate system. The motion due to the actuation is converted into the XY coordinate system because of friction between the robot and the surface. 

2. We need to set the 'phi' value for the direction tracking.  I have considered the constant value of phi =0.1 rad for simplicity. That results in the circular motion of the snake robot due to the constant offsetting head link joint. When the robot head reaches X=0, Y= value, it starts moving back from that. It does not move further in the second quadrant. There is no error or simulation stop due to calculation problems. 
3. Changing of the phi value, direction changing is observed in running simulation. 

The Simulink model and its parameters are uploaded to the repository in a folder named [Robot_head_control](https://github.com/bhavikpatel2/Snake-robot-MATLAB/tree/main/Robot_head_control). If you need some explanation in the video, please inform me. I will make a short video and state the exact problem.

**Reply by robertogl** [reply]
Ok, thanks for the detailed explanation. It may sound like a sign switching or angle wrapping problem, but you said already there is no calculation error. Let me take a look at the model and see if I can find anything.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
Have you found any problems with my model?

**Reply by robertogl** [reply]
I have been trying to understand what the problem is but I am still not too sure. I noticed that the robot suddenly changes its direction but not necessarily at X=0 as you can see in the plot below. I am wondering if that might be due to the limitation of the PD controller and the increased error in the joint angle. Maybe you can add a Ki term and perhaps even have different controllers for each joint. 

I would not mind having a phone call and trying to figure this out together. Let me know.

![image](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/2199339/c7c2bfc8-d687-442d-8bc2-c52f723babac)

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
Thank you so much for your reply. Could we do a phone call after 29th May because nowadays I am on leave?

**Reply by robertogl** [reply]
Sounds good!

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
I am back to work. On which date could we talk to identify the problem and on which platform (Google Meet/ Teams/Zoom)?

**Reply by robertogl** [reply]
Hi @bhavikpatel2,

I apologize for the delay but I had to work on an important deadline. Would you have time tomorrow or Friday?

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
I am fine on Friday. I am following +5:30 GMT. I will connect with you anytime on Teams/ Google Meet/ Zoom whatever suits you.
Email id: bpatel@iitg.ac.in

**Reply by bhavikpatel2** [reply]
Hi @robertogl 
It was a nice conversation with you. 
Finally, I found the problem. The transformation sensor doesn't have full-angle information. It has only absolute value and therefore, the head angle controller failed. I have rectified that problem, and the robot tracks whatever reference angle is given. Finally, I implemented a head controller. You may find the final version of implemented head controller in [Robot_head_control/Snake_Robot_Head_Control.slx].

**Reply by robertogl** [reply]
Awesome! I am so glad to hear that you found and fixed the problem. It sounds like you have done with the project. Is there anything else you are planning to do before submitting the solution? Otherwise, please go ahead and complete the submission form.
Thanks!

**Reply by bhavikpatel2** [reply]
Yes, I will submit it by making readme,md file and comments within two week. 
Thank you.

**Reply by bhavikpatel2** [reply]
Hi @robertogl 

I have submitted the solution and updated the Readme.md file. I tried to make it as simple as possible for learners. If any upgradation is required, kindly suggest me.
Thank you for being so supportive in completing this project.

**Comment by OdlLed234** [comment]
Hi. I am a student at the University of the Witwatersrand, Johannesburg, and interested in working and collaborating on this project

**Comment by Antoine-ms** [comment]
Hi, I am Antoine Masson, a mechanical and control system engineering student at INSA Rennes, France. I am interested in working and collaboring on the project

**Reply by Antoine-ms** [reply]
Hello @robertogl 
I have created a model of the snake robot, but I don't really understand how it is supposed to move forward. Am I supposed to add wheels to it or should it be able to move just by using the friction with the ground ? 
Could you give me some suggestions ? 
Thanks

**Reply by robertogl** [reply]
Hello @Antoine-ms and everyone on this discussion.
Sorry for my silence but I am now back and I will try to answer all your questions. Also, happy new year!

Snake-robot locomotion is indeed not a very easy task. To mimic the actual movement of a snake you should not be using wheels but simulate what is called a "serpentine locomotion". That should simulate the "body dragging" effect of a typical snake motion.
There are several papers online that explain how to implement that locomotion. Here is one that also provides a formula of a "Serpenoid Curve"
https://www.sciencedirect.com/science/article/pii/S2213020915000622

That said, if you find this too difficult or you have limited time, wheels would be a good starting point.

I hope this can help.

**Reply by Antoine-ms** [reply]
Hi @robertogl , 
I implemented the serpentine motion and it is working fine, but my robot is not moving forward. My guess is that the issue comes from the friction model. In order to solve that, I tried to model anisotropic friction with a transform sensor block, but after many tries I couldn't make it work. Is there any chance you can take a look on my issue ? I created a repository with my files on my profile. 
Thanks.

**Reply by robertogl** [reply]
Hi @Antoine-ms,

Friction can be a reason, but I think you should also play around with the internal mechanical parameters of the joints like spring stiffness. For example, In your model I tried changing the spring stiffness and the snake moved even if in the opposite direction :). 
Moreover, you may need that anyway as for what I could see the joint rotations go over the limit of 90deg that you provided.

**Reply by Antoine-ms** [reply]
Hi @robertogl 
Thank you for your answers, I changed the spring stiffness as you suggested and I got my snake to move a little bit, but in the opposite direction. I tried to play around even more with the friction parameters but I couldn't solve this issue, so I created a 2nd version of my snake, with wheels, to confirm that the issue comes from the friction. This new model works very well, my snake moves quite fast and in the good direction this time !
However, this new model is quite slow to compute. Do you have any suggestion to improve the performances of this 2nd model, or to solve the friction issue on the 1st model ?
Thanks

**Reply by robertogl** [reply]
Hi @Antoine-ms,

I checked your first model and now I am even more convinced is a friction issue. I changed your 'friction_x' parameter to a positive value and the robot moves in the proper direction but its velocity increases fast. 
One of the problems might be related to the static friction being too low, in particular in the y direction of each link's body frame. That is why the link slides back laterally when the next join rotates, I think.

**Reply by bhavikpatel2** [reply]
Hi @robertogl @Antoine-ms 

I have followed your conversation and found my previous mistake in the friction model, but my model is not moving forward after updating. When I took the positive value of the friction coefficient, the model behaved as previously. You may check my repository.

**Reply by Antoine-ms** [reply]
Hi @robertogl 
You changed the friction coefficient to a positive value, but doesn't that mean that the force applied is now in the same direction as the velocity ? ie. it is no longer a friction force but a force that is making the snake move forward ? That would explain why the velocity increases fast

**Reply by robertogl** [reply]
Hi @Antoine-ms,
Sorry for the confusion. I was not suggesting to change the coefficient to a positive value but just pointing out that the robot can actually move and thus that probably something in the friction is still wrong. 
I see that you multiply the coefficient with the velocity so that you have always a friction that is in the opposite direction but that is all. Have you tried to add some static friction?

**Reply by Antoine-ms** [reply]
Hi @robertogl 
I searched for a while but I couldn't manage to find how to add static friction when the friction is an external input. Do you have any advice ?

**Reply by robertogl** [reply]
I think you would have to consider that into the same input. Something like, providing a very high friction value until the applied force is higher than a threshold. But I see how it gets complicated. As an alternative for now you can try increasing the normal force as that will have an effect on the friction, if you have not done that yet. Also, have you really found a clear benefit in providing friction by an external input?

**Reply by Antoine-ms** [reply]
Hi @robertogl 
I will try that, thank you !

The purpose of the external input is to have anisotropic friction : a snake robot with isotropic friction can move forward but some papers have demonstrated that it can't be controlled. What's more, the friction along the direction of the movement needs to be as low as possible compared to the friction in the normal direction for the robot to move efficiently. 

I did not find any other way to model anisotropic friction. I also tried without external input by setting the friction coefficients as matrix but I didn't succed and I couldn't find any documentation on this.

Overall, I could not really find any documentation or paper about how to model anisotropic friction in Simulink, do you happen to know if there are some that I did not find ? 
The easiest solution and the one that I saw in some papers is to add wheels to each part of the body, it works really good on my model but the simulation is not anymore in real time so it's very time consuming to wait 5-10 min each time I change a parameter.

**Reply by robertogl** [reply]
I am not aware of any specific documentation or example about anisotropic friction. 
We have this [example](https://www.mathworks.com/matlabcentral/fileexchange/109680-spatial-contact-force-examples-simscape-multibody) about custom friction force. You can take a look, if you have not already, hoping it can help. In the meantime I will ask internally about anisotropic friction in Simscape.

**Reply by Antoine-ms** [reply]
Hi @robertogl 
I am working on the controller for the steering of the snake, but I am struggling to compute the direction where the snake is going. I tried to take the mean of the direction or each link but it does not gives good results. Do you have any suggestion ?

**Reply by robertogl** [reply]
What about calculating it incrementally? If you take the mean of all the directions, taken by the head only within each oscillation period, and then add it to the previous mean, assuming the initial direction is zero or known.  

Quick question, does that mean that you solved the problem of the snake not moving forward or you are using the model with the wheels?

**Reply by Antoine-ms** [reply]
I will try that thanks !

I successfully implemented the anisotropic friction in 3 models : two from the example you gave me that I tweaked in order to get anisotropic friction, and one that is very simplified. 
The 3 models are correctly modelling the anisotropic friction, but the only time I could get the snake moving is when the friction was anisotropic according to the ground frame and not each link. In that case the snake could only move along the global x axis and therefore could not steer. I still don't understand why it won't move forward when the friction is anisotropic according to each link's local frame.

I ended up creating a new and simpler model with only one big wheel per link in order to work on the controller.

**Reply by robertogl** [reply]
Right. The anisotropic friction should be indeed with respect to each link's base frame. In that case Vx is always parallel to the link's longitudinal direction (the longer side of the link) and Vy to the link's lateral direction (the shorted side), if that's clear. I wonder if there might be some error in the reference frame used as base and thus the velocity is not properly computed(?)

**Reply by Antoine-ms** [reply]
Yes that's clear, I'll check my model again, clean it a bit and put it on my repository. I'll let you know when it's done if you want to take a look on the reference frames I used.

**Reply by Antoine-ms** [reply]
I uploaded 3 versions of my anisotropic model on my repository (3_1 ; 3_2 ; 3_3). I checked the reference frames again and I tried to swap them, but I still couldn't manage to get my snake to move.

**Reply by Antoine-ms** [reply]
Hi @robertogl 

I finished the design of my controller, I tested it on several scenario and everything works fine exept for one last problem : 
The desired direction of the snake is computed as an angle in radian between -pi and pi, which raises the problem of the discontinuity. Do you have any idea how I can handle that ? 
Thanks

**Reply by robertogl** [reply]
The error to feed back has to be calculated properly so to identify those jumps and unwrap their values when bigger than a certain threshold ('pi') by adding 2pi with the appropriate sign. For these cases our [unwrap](https://www.mathworks.com/help/matlab/ref/unwrap.html) function will be useful.

**Reply by Antoine-ms** [reply]
Hi @robertogl 

I am struggling to get the unwrap block to work. It seems like it has no effect on the signal. I tried to play with its parameters but I couldn't make it work. Do you think you could take a look on that ?
Thanks

**Reply by robertogl** [reply]
I am having some issues in running the model. But have you tried applying the unwrap only to the error angle?

**Comment by Alexan41** [comment]
Is the competition still on?

**Reply by robertogl** [reply]
Hi @Alexan41,

Currently the 'Sustainability and Renewable Energy' challenge is on until January. However, this project does not qualify as part of that. But if you mean this project, yes this is still active!

Cheers,
Roberto

**Reply by Alexan41** [reply]
Okay but I focus on the snake like robot project because I've already started working on it.

**Reply by Alexan41** [reply]
Please @robertogl , 

I made progress with my project and I want to share with you what I have done so far so that you can appreciate. 
I don't know very weel how github works so I would need some tips to share my project with you.

Thanks in advance.

**Reply by robertogl** [reply]
Hi @Alexan41,

You can easily use this [template](https://github.com/robertogl/ProjectSolution). Just click on 'Use this template' and then 'Create your repository'. Once you have done that you just need to change the name of the repository and the owner (you will see your profile name in the drop down menu), and then just click on 'submit'. In the new repository you can literally drag and drop the files you need to share. 

That's a good start to share your initial files. Let me know if you need anything else.

**Reply by Alexan41** [reply]
It is done. I have submitted my work.
Thank you Sir.

**Reply by robertogl** [reply]
Thanks @Alexan41! I have received the submission, but I am bit behind with reviews because of other commitments. I will get back to you as soon as possible. The video looks great btw 🙂!

**Reply by Alexan41** [reply]
Thank you sir!

**Reply by Alexan41** [reply]
Hi @robertogl
I think you forgot me, I'm still waiting for feedback on my project.

**Comment by RKommX** [comment]
Hi, I'm Rajdeep Paul, 2nd year aerospace engineering student at IIT Bombay. I wish to spend my semester break working on this project.

**Reply by RKommX** [reply]
Hi @robertogl

I've begun work on the Simscape model. I've made a 3-link snake and a test wave generator for now before I move on to modelling an autonomous control system. The problem is that it is taking too long to simulate (almost 1 ms per 3 s). I tried changing the solver and using the accelerator, but they didn't work.

I have attached the link to my repo here. Please review it and let me know if anything is amiss. I'm fairly new to Simscape, so I'd take any help to learn from my mistakes.

Thank you.
[https://github.com/RKommX/Snake_Robot](url)

**Reply by RKommX** [reply]
Hey @robertogl 

I figured the spatial contact forces may have been creating problems, so I removed them and used planar joints instead of 6-DOF joints. I also used a LOS control for the robot. The simulation runs to completion, but the robot shoots away, so I think there is some unstability in the system.

I've updated my repo. Please take a look if you have time and let me know what can be done.

[1-s2.0-S1367578817301050-main (1).pdf](https://github.com/user-attachments/files/18094584/1-s2.0-S1367578817301050-main.1.pdf)
This is the source I'm referring to.
Thanks.

**Reply by robertogl** [reply]
Hi @RKommX,

I have tried to run your model but I got several errors. I fixed a coupled but not all apparently. Can you check if your model works fine after you download it from the repo in a fresh folder?

Also, I suggest you to take a look at the Point Cloud block or the Sphere block to model the geometry of each link of the snake that makes contact with the ground. Each link would have a point cloud associated with it, and then that point cloud has a contact force block between it and the ground. Then tune the parameters appropriately. You might be able to get away using a sphere instead of a point cloud. Here is a decent [example](https://www.mathworks.com/help/sm/ug/grid_surface_ball.html).

Another option is taking a look at Bhavik's submission. Perhaps you can build your solution on top of that and create a more complete simulation of an autonomous snake robot that includes sensors and autonomous navigation in a more complex environment. Up to you.

**Reply by RKommX** [reply]
Hi @robertogl,

I finally got it to work. I changed the gain value from the LOS block and a few other parameters. It moves smoothly, but the only issue is that it seems to move slightly downwards (very slight compared to its forward motion). I have updated my repo, so please check if it's alright.

I do plan on making an underwater version. I will also try a sphere link. For now, I want to brush this up and try to reduce its deviation as much as possible.

Thank you.

Edit: Reattaching the link, previous one gave 404
[https://github.com/RKommX/Snake_Robot](url)

**Reply by robertogl** [reply]
Hi @RKommX, sorry for the delay. Please allow me a few more days and get back to you.

**Reply by RKommX** [reply]
No issues. I've already been bogged down by academics, so I haven't been able to work on it for a while.
